Dataflux

uriel-guzman opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

amacaskill commented on pkg/scanner/scanner.go:
```diff
@@ -600,48 +650,212 @@ func (s *Scanner) syncPV(ctx context.Context, key string) error {
 	duration := timeNow().Sub(syncStartTime)
-	s.eventRecorder.Eventf(pv, v1.EventTypeNormal, reasonScanOperationSucceeded, "Bucket scan completed successfully for bucket %s, directory '%s' (%v)", bucketInfo.name, bucketInfo.dir, duration)
-	if patchErr := s.updatePVScanResult(ctx, pv, info, scanCompleted); patchErr != nil {
+	s.eventRecorder.Eventf(pv, v1.EventTypeNormal, reasonScanOperationSucceeded, "Bucket scan completed successfully for bucket %s, directory '%s' (%v): %d objects, %d bytes", bucketInfo.name, bucketInfo.dir, duration.Round(time.Second), bucketInfo.numObjects, bucketInfo.totalSizeBytes)
```
'%s' -> %q

%q will put quotes around the string

amacaskill replied:
Ok, we can always make changes like that in follow up PRs if its going to be that big of change. Let's leave the change in since you already made it, but in the future

uriel-guzman replied:
I already updated the PR to change `%s` to `%q`. So I think we can leave this in now.

amacaskill commented on pkg/scanner/scanner.go:
```diff
@@ -565,7 +615,7 @@ func (s *Scanner) syncPV(ctx context.Context, key string) error {
 		return nil // Remove irrelevant PV from queue
 	}
-	klog.Infof("PV %q is relevant, bucket: %s, dir: %s", key, bucketInfo.name, bucketInfo.dir)
+	klog.Infof("PV %q is relevant, bucket: %s, dir: %s, onlyDirSpecified: %v", key, bucketInfo.name, bucketInfo.dir, bucketInfo.onlyDirSpecified)
```
here and throughout code where you use bool in infof

uriel-guzman replied:
Fixed.

uriel-guzman commented on cmd/csi_driver/main.go:
```diff
@@ -51,6 +51,9 @@ var (
 	identityProvider             = flag.String("identity-provider", "", "The Identity Provider to authenticate with GCS API.")
 	enableProfiling              = flag.Bool("enable-profiling", false, "enable the golang pprof at port 6060")
 	enableBucketScanner          = flag.Bool("enable-bucket-scanner", false, "enable the bucket scanner feature")
+	datafluxParallelism          = flag.Int("dataflux-parallelism", 0, "number of go routines for Dataflux lister. Defaults to 0 (10X number of available vCPUs)")
```
`0` in the Dataflux package automatically assigns the parallelism to be 10X the number of the available CPU: https://github.com/googleapis/google-cloud-go/blob/storage/v1.56.1/storage/dataflux/fast_list.go#L75. This is the default behavior.

amacaskill commented on pkg/scanner/scanner.go:
```diff
@@ -300,6 +334,22 @@ func (s *Scanner) Run(ctx context.Context) {
 	defer runtime.HandleCrash()
 	defer s.queue.ShutDown()
 
+	gcsClient, err := storage.NewClient(ctx)
+	if err != nil {
+		klog.Errorf("Failed to create GCS client: %v", err)
+		return
+	}
+	s.gcsClient = gcsClient
+	defer gcsClient.Close()
+
+	metricClient, err := monitoring.NewMetricClient(ctx)
+	if err != nil {
+		klog.Errorf("Failed to create metric client: %v", err)
+		return
+	}
+	s.metricClient = metricClient
+	defer metricClient.Close()
+
```
Why don't we initialize these clients in NewScanner? Similar to other clients like kubeClient?

amacaskill replied:
Yea the only difference is we hit the error during driver initialization rather than later during Run, but I think there's pros and cons to both.

amacaskill commented on pkg/scanner/scanner.go:
```diff
@@ -600,48 +650,212 @@ func (s *Scanner) syncPV(ctx context.Context, key string) error {
+		// Ensure that the directory name ends with "/" to avoid picking up
+		// files with prefixes of other directories, since GCS "nested"
+		// objects are just file names with "/" in the names.
+		dfInput.Query.Prefix = strings.Trim(info.dir, "/") + "/"
```
Please make sure you have e2e tests for this case in future PR

uriel-guzman replied:
Added TODO here.

amacaskill commented on pkg/scanner/scanner.go:
```diff
@@ -600,48 +650,212 @@ func (s *Scanner) syncPV(ctx context.Context, key string) error {
+}
+
+// fetchMetricValue retrieves the latest value for a given metric type from Cloud Monitoring.
+func fetchMetricValue(ctx context.Context, metricClient *monitoring.MetricClient, projectNumber string, info *bucketInfo, metricType string) (int64, error) {
```
I'm curious of performance of this. Whats the max number of times we can call this per day? Is there a way to cut down on number of calls / AlignmentPeriod if we detect it was called recently?

uriel-guzman replied:
This is only being called once per scan, which by default is once per week and can be configured by the customer (if they want, but they normally shouldn't). The scan is limited by the `bucketScanTTL` parameter, so it will not be called more than once per that time window.

amacaskill replied:
Do we have a check in place somewhere/test that will fail if buckets start being scanned more than once per week for some reason (or more than bucketScanTTL)?

uriel-guzman replied:
Yes, the basic gcsfuse storage profiles test will have a test case where a Pod is created soon after the first scan, before the `bucketScanTTL`, and will explicitly verify the timestamp to make sure the test is not triggered. Subsequently, another Pod will be created after the `bucketScanTTL` expires and verify that the scan is triggered.

Additionally, there are already 2 unit tests that explicitly verify this. See https://github.com/GoogleCloudPlatform/gcs-fuse-csi-driver/blob/a453b1824f646b6425e4ec938cd1dd148920fd4d/pkg/scanner/scanner_test.go#L411

uriel-guzman commented on examples/gcsfuse-profiles/README.md:
````diff
@@ -0,0 +1,60 @@
+```bash
+GCS_PROJECT=your-bucket-project-id
+BUCKET_NAME=your-bucket-name
+PROJECT_NUMBER=$(gcloud projects describe $(gcloud config get project) --format="value(projectNumber)")
````
Resolved.

uriel-guzman commented on pkg/scanner/scanner.go:
```diff
@@ -600,48 +650,212 @@ func (s *Scanner) syncPV(ctx context.Context, key string) error {
+	klog.Infof("Bucket %s HNS enabled: %v", info.name, info.isHNSEnabled)
+
+	if info.onlyDirSpecified {
+		klog.Infof("onlyDirSpecified is true for bucket %s, dir '%s'. Scanning with Dataflux.", info.name, info.dir)
```
Resolved.

uriel-guzman commented on pkg/scanner/scanner.go:
```diff
@@ -600,48 +650,212 @@ func (s *Scanner) syncPV(ctx context.Context, key string) error {
+			return nil
+		}
+
+		s.eventRecorder.Eventf(pv, v1.EventTypeWarning, reasonScanOperationWarning, "Failed to scan bucket %s using GCS Bucket Metrics, falling back to Dataflux: %v", info.name, mErr)
```
Fixed.

amacaskill commented on pkg/scanner/scanner.go:
```diff
@@ -600,48 +650,212 @@ func (s *Scanner) syncPV(ctx context.Context, key string) error {
+// unavailable metrics.
+//
+// Optionally, GCS Dataflux client scanning on the entire bucket can be forced by specifying `only-dir=/`
+func defaultScanBucket(s *Scanner, ctx context.Context, info *bucketInfo, scanTimeout time.Duration, pv *v1.PersistentVolume) error {
```
All the Nits are optional, you can choose to reject them if you want

uriel-guzman replied:
Fixed. Note that this was used all over the place, so there's going to be several hundreds of lines for this change as well, since it's a big refactoring operation.

uriel-guzman commented:
> What did you change between the tests: `if only-dir not specified and bucket metrics fail` and `only-dir not specified and bucket metrics succeed`. The output looks the same?
> 
> Also did you test `only-dir=/` ? is that covered by only-dir tests in description

Yeah, that was not right. Updated that test (it shouldn't show any warning) and added the `only-dir=/` test as #6.
